crates/dekaf/src/task_manager.rs

```diff
@@ -0,0 +1,450 @@
+
+                let _ = self
+                    .update_partition_info(
+                        &task_name,
+                        &spec,
+                        &mut partitions,
+                        journal_clients.clone(),
+                    )
+                    .await?;
+
+                let _ = sender.send(Some(Ok(TaskState {
+                    access_token,
+                    access_token_claims,
+                    ops_logs_journal,
+                    ops_stats_journal,
+                    spec,
+                    partitions: partitions
+                        .iter()
+                        .sorted_by_key(|(k, _)| k.as_str())
+                        .map(|(k, v)| {
+                            let mapped_val = match v {
+                                Ok(p) => Ok(p.clone()),
+                                Err(e) => Err(TaskManagerError::from(e)),
+                            };
+                            let res = (k.clone(), mapped_val);
+
+                            res
+                        })
+                        .collect_vec(),
+                })));
+
+                Ok(())
+            }
+            .await;
+
+            if let Err(e) = loop_result {
+                tracing::error!(task_name, error=%e, "Error in task manager loop");
+                let _ = sender.send(Some(Err(TaskManagerError::from(e))));
+            }
+
+            tokio::select! {
+                _ = stop_signal.cancelled() => {
+                    tracing::info!(task_name, "signalled to stop");
+                    break;
+                }
+                _ = interval.tick() => {}
+            }
+        }
+        drop(receiver);
+
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(self, spec, partitions, journal_clients))]
+    async fn update_partition_info(
+        self: &std::sync::Arc<Self>,
+        task_name: &str,
+        spec: &MaterializationSpec,
+        partitions: &mut HashMap<String, anyhow::Result<Vec<Partition>>>,
+        journal_clients: Arc<
+            std::sync::Mutex<HashMap<String, (journal::Client, proto_gazette::Claims)>>,
+        >,
+    ) -> anyhow::Result<()> {
+        let mut current_partition_template_names = HashSet::with_capacity(spec.bindings.len());
+        let mut tasks = Vec::with_capacity(spec.bindings.len());
+
+        for binding in &spec.bindings {
+            let collection_spec = binding
+                .collection
+                .as_ref()
+                .context("expected collection Spec")?;
+            let partition_template = collection_spec
+                .partition_template
+                .as_ref()
+                .context("expected partition template")?;
+
+            let partition_selector = binding
+                .partition_selector
+                .as_ref()
+                .context("expected partition selector")?;
+
+            current_partition_template_names.insert(partition_template.name.clone());
```

Comment:
why do you need this? just gather from `tasks` on the way out?

Reply:
Yeah... it's because of [this](https://github.com/estuary/flow/pull/2116#discussion_r2070526286). I like the scatter/gather approach better.